File: Programmers/[42579] 베스트앨범.py
from collections import deque


def solution(genres, plays):
    answer = []
    playlist = dict()
    genres_total = dict()

    for i, (genre, play) in enumerate(zip(genres, plays)):
        if genre in playlist:
            playlist[genre].append((play, i))
            genres_total[genre] += play
        else:
            playlist[genre] = [(play, i)]
            genres_total[genre] = play

    for pl in playlist:
        playlist[pl] = sorted(list(playlist[pl]), key=lambda x: (-x[0], x[1]))[:2]
    sorted_genres_total = sorted(genres_total.items(), key=lambda x: x[1], reverse=True)

    for genre in sorted_genres_total:
        for i in playlist[genre[0]]:
            answer.append(i[1])

    return answer


if __name__ == "__main__":
    # genresm, plays = ["classic", "pop", "classic", "classic", "pop"], [500, 600, 150, 800, 2500]
    genresm, plays = ["classic", "SS", "SS", "SS", "SS", "pop"], [6000, 1, 4, 4, 4, 2500]
    print(solution(genresm, plays))

File: Programmers/[42579] 베스트앨범.java
package pg42579;

import java.util.*;

class Solution {
    static Map<String, Integer> genreMap = new HashMap<>();

    public int[] solution(String[] genres, int[] plays) {
        List<Music> musics = new ArrayList<>();
        for (int i = 0; i < genres.length; i++) {
            Integer genrePlayedCnt = genreMap.get(genres[i]);
            genreMap.put(genres[i], genrePlayedCnt == null ? plays[i] : genrePlayedCnt + plays[i]);

            musics.add(new Music(genres[i], i, plays[i]));
        }

        Collections.sort(musics);

        List<Integer> answer = new ArrayList<>();
        genreMap.replaceAll((key, value) -> 0);

        for (Music music : musics) {
            Integer curGenreCnt = genreMap.get(music.genre);
            if (curGenreCnt < 2) {
                genreMap.put(music.genre, curGenreCnt + 1);
                answer.add(music.idx);
            }
        }

        return answer.stream().mapToInt(x -> x).toArray();
    }

    static class Music implements Comparable<Music> {
        String genre;
        int idx;
        int playedCnt;

        public Music(String genre, int idx, int playedCnt) {
            this.genre = genre;
            this.idx = idx;
            this.playedCnt = playedCnt;
        }

        @Override
        public int compareTo(Music o) {
            if (!genreMap.get(this.genre).equals(genreMap.get(o.genre))) {
                return genreMap.get(o.genre) - genreMap.get(this.genre);
            }
            if (this.playedCnt == o.playedCnt) return this.idx - o.idx;
            return o.playedCnt - this.playedCnt;
        }
    }
}